File: practice/groupAnagrams.go
/* Group Anagrams
Given an array of strings, group anagrams together.
*/
package main

import (
	"sort"
	"strings"
	"fmt"
)

func groupAnagrams(strs []string) [][]string {
	anagramsMap := make(map[string][]string)
	for _, str:= range strs{
		sortedStr := strings.Split(str, "")
		sort.Strings(sortedStr)
		key := strings.Join(sortedStr, "")
		anagramsMap[key] = append(anagramsMap[key], str)
	}
	result := [][]string{}
	for _, group := range anagramsMap{
		result = append(result, group)
	}
	return result
}

func main() {
	strs := []string{"eat", "tea", "tan", "ate", "nat", "bat"}
	anagrams := groupAnagrams(strs)
	for _, group := range anagrams {
		fmt.Println(group)
	}

}

File: practice/groupAnagrams.py
'''
Group Anagrams
Given an array of strings, group anagrams together.
'''
from collections import defaultdict
from typing import List
def group_anagrams(strs: List[str]) -> List[List[str]]:
    """
    Groups anagrams from the input list of strings.
    Time Complexity: O(n * k log k) where n is the number of strings and k is the maximum length of a string.
    Space Complexity: O(n) for storing the grouped anagrams.
    """
    anagrams = defaultdict(list)
    for s in strs:
        sorted_str = ''.join(sorted(s)) # Sort the string to find its anagram key
        anagrams[sorted_str].append(s)
    return list(anagrams.values())
if __name__ == "__main__":
    test_cases = [
        ["eat", "tea", "tan", "ate", "nat", "bat"],
        [""],
        ["a"],
        ["abc", "bca", "cab", "xyz", "zyx"]
    ]
    for strs in test_cases:
        print(f"Input: {strs}, Grouped Anagrams: {group_anagrams(strs)}")